### PR TITLE
Create backup restore e2e test for service-catalog domain

### PR DESCRIPTION
**Pull request by polskikiel**
**Description**

Changes proposed in this pull request:

**Created backup restore e2e test for:**
- Service Catalog flow
- App broker 
- Helm broker

**How to test**
1. Run cluster with kyma
2. Create ark deployment (with secrets to azure)
3. Create telepresence deployment
```
kubectl run myservice --image=datawire/telepresence-k8s:0.97 --port=8000 --expose
```

4. Run test using this command: 
```
env KUBECONFIG={kubeconfig_path_to_cluster_with_ark_setup} telepresence --swap-deployment myservice --expose 8080 --run  go test -count=1 -timeout=45m
``` 

**Example output**
Screenshot 2019-03-22 at 10 25 28
Screenshot 2019-03-22 at 10 25 56

**Related issue(s)**

### CODE REVIEW

File: tests/end-to-end/backup-restore-test/.gitignore [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+*.iml
+e2e-upgrade-test
+/vendor
+
+# This is ignored on purpose.
+# The `installation/scripts/e2e-testing.sh` script is traversing through end-to-end directory and checks if
+# sub-project has deploy folder with chart definition, if yes then this chart is installed.
+# We do not want to have this behaviour because our `upgrade` chart should be used only on upgrade ProwJob plan
+/deploy
```
Comment (mszostok) [inline tests/end-to-end/backup-restore-test/.gitignore]: you cannot have that here 😂 Here this behavior is expected one.

File: tests/end-to-end/backup-restore-test/.gitignore [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+*.test
+
+# Output of the go coverage tool, specifically when used with LiteIDE
+*.out
+
+# Project-local glide cache, RE: https://github.com/Masterminds/glide/issues/736
+.glide/
+
+*.iml
+e2e-upgrade-test
```
Comment (mszostok) [inline tests/end-to-end/backup-restore-test/.gitignore]: this should be name of generated binary,

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/brokers_flow.go [hunk 3]
```diff
@@ -0,0 +1,372 @@
+package service_catalog
+
+import (
+	"github.com/sirupsen/logrus"
```
Comment (mszostok) [inline tests/end-to-end/backup-restore-test/backupe2e/service-catalog/brokers_flow.go]: sort import, right now you have groups that don't make sense


same suggestion for rest

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+			f.logReport()
+		}
+		So(err, ShouldBeNil)
+	}
+}
+
+func (f *helmBrokerFlow) testResources() {
+	// iterate over testing steps
+	for _, fn := range []func() error{
+		f.storeKeyInRedis,
```
Comment (mszostok) [inline tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go]: as discussed, we should check if value exists there, 

additionaly, you can try to store new entry to check if connection is still working properly.

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/app_broker.go [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+		}
+		So(err, ShouldBeNil)
+	}
+}
+
+func (f *appBrokerFlow) testResources() {
+	// iterate over testing steps
+	for _, fn := range []func() error{
+		f.verifyEventActivation,
+		f.verifyEnvTesterHasGatewayInjected,
```
Comment (mszostok) [inline tests/end-to-end/backup-restore-test/backupe2e/service-catalog/app_broker.go]: consider adding sbu deletion and check if env was removed

File: tests/end-to-end/backup-restore-test/.gitignore (hunk 1 above)
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/.gitignore]: I would probably find this highly copied .gitignore during image bump ;p

File: tests/end-to-end/backup-restore-test/Makefile [hunk 6]
```diff
@@ -9,8 +9,8 @@ build:
 .PHONY: build-image
 build-image:
 	mkdir -p build-temp
-	cp ../../../docs/backup/docs/assets/all-backup.yaml build-temp
-	cp ../../../docs/backup/docs/assets/system-backup.yaml build-temp
+	cp all-backup.yaml build-temp
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/Makefile]: reverted

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go (hunk 4 above)
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go]: We are checking if the connection is still working properly after restore by getting value from it so I decided not to add another assertion.

File: tests/end-to-end/backup-restore-test/restore_cluster_backup_test.go [hunk 7]
```diff
@@ -97,13 +112,15 @@ func TestBackupAndRestoreCluster(t *testing.T) {
...
-	Convey("Create resources", t, func() {
+	Convey("Create resources\n", t, func() {
 		for _, e2eTest := range e2eTests {
+			log.Infof("Creating Namespace: %s", e2eTest.namespace)
 			err := myBackupClient.CreateNamespace(e2eTest.namespace)
 			So(err, ShouldBeNil)
 			e2eTest.backupTest.CreateResources(e2eTest.namespace)
 		}
 		for _, e2eTest := range e2eTests {
+			log.Infof("Testing resources in namespace: %s", e2eTest.namespace)
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/restore_cluster_backup_test.go]: The word `Create` is already much abused in logs so I prefer start this one from `Testing` to distinguish it from the other logs.

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+func (f *helmBrokerFlow) testResources() {
+	for _, fn := range []func() error{
+		f.waitForRedisInstance,
+		f.verifyKeyInRedisExists,
+		f.verifyDeploymentContainsRedisEvns,
+		f.deleteRedisBindingUsage,
+		f.verifyDeploymentDoesNotContainRedisEnvs,
+		f.createRedisBindingUsage,
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go]: done

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+	client, err := f.redisClient()
+	if err != nil {
+		return err
+	}
+
+	err = f.wait(time.Minute, func() (done bool, err error) {
+		if client.Ping().Val() == "PONG" {
+			return true, nil
+		}
+		f.log.Info("Redis does not answer. Retry.")
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go]: added

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+		return false, nil
+	})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go]: done

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+func (f *helmBrokerFlow) verifyKeyInRedisExists() error {
+	f.log.Info("Verify the value stored in the Redis DB")
+
+	client, err := f.redisClient()
+	if err != nil {
+		return err
+	}
+
+	// wait is required after restore
+	f.wait(time.Minute*2, func() (done bool, err error) {
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go]: done

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+func (f *helmBrokerFlow) deleteRedisBinding() error {
+	return f.deleteServiceBinding(redisBindingName)
+}
+
+func (f *helmBrokerFlow) deleteRedisInstance() error {
+	return f.deleteServiceInstance(redisInstanceName)
+}
+
+// that is necessary to restore correctly the etcd data
+// works only for single-member etcd cluster
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/backupe2e/service-catalog/helm_broker.go]: added the assertion to check that: 
```
if etcd.Spec.Replicas == nil || *etcd.Spec.Replicas > 1 {
  return errors.Errorf("etcd has many members")
}
```

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/app_broker.go [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+func (f *appBrokerFlow) testResources() {
+	for _, fn := range []func() error{
+		f.verifyApplication,
+		f.waitForClassAndPlans,
+		f.waitForAppInstances,
+		f.verifyEventActivation,
+		f.verifyEnvTesterHasGatewayInjected,
+		f.deleteAPIServiceBindingUsage,
+		f.verifyEnvTesterHasGatewayNotInjected,
+		f.createAPIServiceBindingUsage,
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/backupe2e/service-catalog/app_broker.go]: to restore it after the tests, added a comment

File: tests/end-to-end/backup-restore-test/backupe2e/service-catalog/app_broker.go [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+	err := f.wait(2*time.Minute, func() (bool, error) {
+		expectedClasses := map[string]struct{}{
+			apiServiceId:    {},
+			eventsServiceId: {},
+		}
+		classes, err := f.scInterface.ServicecatalogV1beta1().ServiceClasses(f.namespace).List(metav1.ListOptions{})
+		if err != nil {
+			return false, err
+		}
+		for _, class := range classes.Items {
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/backupe2e/service-catalog/app_broker.go]: done

File: tests/end-to-end/backup-restore-test/restore_cluster_backup_test.go [hunk 15]
```diff
@@ -68,6 +80,9 @@ func TestBackupAndRestoreCluster(t *testing.T) {
 		myFunctionTest,
 		myDeploymentTest,
 		myStatefulSetTest,
+		myHelmBrokerTest,
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/restore_cluster_backup_test.go]: no it's not, changed.

File: tests/end-to-end/backup-restore-test/restore_cluster_backup_test.go [hunk 16]
```diff
@@ -115,42 +132,47 @@ func TestBackupAndRestoreCluster(t *testing.T) {
 		systemBackupSpecFile := cfg.SystemBackupConfigurationFile
 		systemBackupName := "system-" + backupName
 
-		err := myBackupClient.CreateBackup(allBackupName, allBackupSpecFile)
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply tests/end-to-end/backup-restore-test/restore_cluster_backup_test.go]: https://github.com/kyma-project/kyma/pull/3174#pullrequestreview-221123481
https://github.com/kyma-project/kyma/pull/3218#pullrequestreview-217697356
Authors of the test framework clearly don't want to change it.

**Pull request merged**